classes/product/SpecificPriceFormatter.php

```diff
@@ -0,0 +1,179 @@
+ * This code has been extracted here for unit testing purpose
+ *
+ * @todo Refactor this class
+ */
+class SpecificPriceFormatterCore
+{
+    /**
+     * Calculation method to be used (tax included or not?)
+     *
+     * @var bool
+     */
+    private $isTaxIncluded;
+
+    /**
+     * Specific price data array
+     *
+     * @var array
+     */
+    private $specificPrice;
+
+    /**
+     * @var Context
+     */
+    private $context;
+
+    /**
+     * SpecificPriceFormatter constructor.
+     *
+     * @param array $specificPrice
+     * @param bool $isTaxIncluded
+     * @param Context $context
+     */
+    public function __construct(array $specificPrice, bool $isTaxIncluded, Context $context)
+    {
+        $this->specificPrice = $specificPrice;
+        $this->isTaxIncluded = $isTaxIncluded;
+        $this->context = $context;
+    }
+
+    /**
+     * This is legacy code extracted from ProductController and it should be refactored
+     *
+     * @param float $initialPrice
+     * @param float $tax_rate
+     * @param float $ecotax_amount
+     *
+     * @return array
+     */
+    public function formatSpecificPrice($initialPrice, $tax_rate, $ecotax_amount)
+    {
+        $priceFormatter = new PriceFormatter();
+
+        $this->specificPrice['quantity'] = &$this->specificPrice['from_quantity'];
+        if ($this->specificPrice['price'] >= 0) {
+            // The price may be directly set
+
+            /** @var float $currentPriceDefaultCurrency current price with taxes in default currency */
+            $currentPriceDefaultCurrency = $this->calculateSpecificPrice(
+                $this->specificPrice['price'],
+                (bool) $this->specificPrice['reduction_tax'],
+                $tax_rate,
+                $ecotax_amount
+            );
+
+            // Since this price is set in default currency,
+            // we need to convert it into current currency
+            $this->specificPrice['id_currency'];
+            $currentPriceCurrentCurrency = \Tools::convertPrice($currentPriceDefaultCurrency, $this->context->currency, true, $this->context);
+
+            if ($this->specificPrice['reduction_type'] == 'amount') {
+                $currentPriceCurrentCurrency -= ($this->specificPrice['reduction_tax'] ? $this->specificPrice['reduction'] : $this->specificPrice['reduction'] / (1 + $tax_rate / 100));
+                $this->specificPrice['reduction_with_tax'] = $this->specificPrice['reduction_tax'] ? $this->specificPrice['reduction'] : $this->specificPrice['reduction'] / (1 + $tax_rate / 100);
+            } else {
+                $currentPriceCurrentCurrency *= 1 - $this->specificPrice['reduction'];
+            }
+            $this->specificPrice['real_value'] = $initialPrice > 0 ? $initialPrice - $currentPriceCurrentCurrency : $currentPriceCurrentCurrency;
+            $discountPrice = $initialPrice - $this->specificPrice['real_value'];
+
+            if (Configuration::get('PS_DISPLAY_DISCOUNT_PRICE')) {
```

Comment:
Maybe you can inject this configuration in the constructor as well?
So that we limit as much as possible the references to other legacy classes

Reply:
@jolelievre  done !